### PR TITLE
Option to disable gardener shoot monitoring

### PR DESCRIPTION
**Pull request by nschad**
**How to categorize this PR?**

/area monitoring
/kind api-change
/kind enhancement

**What this PR does / why we need it**:

See Discussion: https://github.com/gardener/gardener/issues/5650
TL;DR Work on the monitoring will not be continued until further notice. Since you can roll your own implementation with a gardener extension one should be able to disable the gardener monitoring.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature operator
Add option to disable gardener shoot monitoring
```

### CODE REVIEW

File: pkg/operation/botanist/monitoring.go [hunk 1]
```diff
@@ -69,6 +69,11 @@ func observabilityIngressSecretConfig(name string) *secrets.BasicAuthSecretConfi
...
 // to monitor the Shoot cluster whose control plane runs in the Seed cluster.
 func (b *Botanist) DeploySeedMonitoring(ctx context.Context) error {
+	// disable monitoring set if disabled in gardenletConfig
+	if !gardenlethelper.IsMonitoringEnabled(b.Config) {
+		return b.DeleteSeedMonitoring(ctx)
+	}
+
 	if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting {
 		return b.DeleteSeedMonitoring(ctx)
 	}
```
Comment (shafeeqes) [inline pkg/operation/botanist/monitoring.go]: Nit: you could combine the two if's.

File: pkg/operation/botanist/monitoring.go [hunk 2]
```diff
@@ -424,6 +429,11 @@ func (b *Botanist) DeploySeedMonitoring(ctx context.Context) error {
 
 // DeploySeedGrafana deploys the grafana charts to the Seed cluster.
 func (b *Botanist) DeploySeedGrafana(ctx context.Context) error {
+	// disable monitoring set if disabled in gardenletConfig
+	if !gardenlethelper.IsMonitoringEnabled(b.Config) {
+		return b.DeleteGrafana(ctx)
```
Comment (shafeeqes) [inline pkg/operation/botanist/monitoring.go]: Shouldn't we cleanup the monitoring secrets as well? Ref:
https://github.com/gardener/gardener/blob/3d001a092ee093ee579ac1d30ae38e3cca107c3e/pkg/operation/botanist/monitoring.go#L427-L434

Maybe you can combine the two if's?

**Review by shafeeqes — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/apis/config/helper/helpers.go [hunk 3]
```diff
@@ -109,3 +109,12 @@ func IsLokiEnabled(c *config.GardenletConfiguration) bool {
 	}
 	return true
 }
+
+// IsMonitoringEnabled return true if the monitoring stack for clusters is enabled. Default is enabled.
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/helper/helpers.go]:
```suggestion
// IsMonitoringEnabled returns true if the monitoring stack for shoot clusters is enabled. Default is enabled.
```

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 4]
```diff
@@ -681,6 +681,9 @@ type MonitoringConfig struct {
 
 // ShootMonitoringConfig contains settings for the shoot monitoring stack.
 type ShootMonitoringConfig struct {
+	// Enabled is used to enable or disable the shoot monitoring stack. If not set the monitoring stack is enabled.
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/v1alpha1/types.go]:
```suggestion
// Enabled is used to enable or disable the shoot monitoring stack.
// Defaults to true.
```

File: pkg/gardenlet/apis/config/helper/helpers.go [hunk 5]
```diff
@@ -109,3 +109,12 @@ func IsLokiEnabled(c *config.GardenletConfiguration) bool {
...
+func IsMonitoringEnabled(c *config.GardenletConfiguration) bool {
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/helper/helpers.go:114]: Can you add a unit test for this function?

File: docs/monitoring/README.md [hunk 6]
```diff
@@ -69,4 +69,13 @@ monitoring:
       additional: label
 ```
 
-If basic auth is needed it can be set via secret in garden namespace (Gardener API Server). [Example secret](../../example/10-secret-remote-write.yaml)
\ No newline at end of file
+If basic auth is needed it can be set via secret in garden namespace (Gardener API Server). [Example secret](../../example/10-secret-remote-write.yaml)
+
+
```
Comment (plkokanov) [inline docs/monitoring/README.md]:
```suggestion
```

File: pkg/gardenlet/apis/config/types.go [hunk 7]
```diff
@@ -562,6 +562,8 @@ type MonitoringConfig struct {
 
 // ShootMonitoringConfig contains settings for the shoot monitoring stack.
 type ShootMonitoringConfig struct {
+	// Enabled is used to enable or disable the shoot monitoring stack. If not set the monitoring stack is enabled.
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/types.go]:
```suggestion
	// Enabled is used to enable or disable the shoot monitoring stack.
	// Defaults to true.
```

File: pkg/operation/botanist/monitoring.go (hunk 2 above)
Comment (plkokanov) [reply pkg/operation/botanist/monitoring.go]: Hmm, don't we use grafana also for loki and logging?

File: pkg/operation/care/health_check_test.go [hunk 8]
```diff
@@ -774,7 +774,7 @@ var _ = Describe("health check", func() {
 				checker           = care.NewHealthChecker(map[gardencorev1beta1.ConditionType]time.Duration{}, nil, nil, kubernetesVersion, gardenerVersion)
 			)
 
-			exitCondition, err := checker.CheckMonitoringControlPlane(seedNamespace, isTestingShoot, wantsAlertmanager, condition, deploymentLister, statefulSetLister)
+			exitCondition, err := checker.CheckMonitoringControlPlane(seedNamespace, isTestingShoot, true, wantsAlertmanager, condition, deploymentLister, statefulSetLister)
```
Comment (plkokanov) [inline pkg/operation/care/health_check_test.go]: Can you also add a test where `monitoringEnabled` is false?

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/monitoring.go (hunk 2 above)
Comment (nschad) [reply pkg/operation/botanist/monitoring.go:434]: yep @plkokanov but that is unchanged. `monitoring` is really just prometheus metrics as far as we can tell

Comment (nschad) [reply pkg/operation/botanist/monitoring.go]: > Hmm, don't we use grafana also for loki and logging?

valid point @plkokanov. We don't need it therefore it was removed but I don't really have a strong opinion about it, whatever seems good for you guys :)

File: pkg/gardenlet/apis/config/helper/helpers.go (hunk 5 above)
Comment (nschad) [reply pkg/gardenlet/apis/config/helper/helpers.go:114]: done!

File: pkg/operation/care/health_check_test.go (hunk 8 above)
Comment (nschad) [reply pkg/operation/care/health_check_test.go]: done! Please check if that is sufficient

File: pkg/gardenlet/apis/config/helper/helpers_test.go [hunk 9]
```diff
@@ -121,6 +121,35 @@ var _ = Describe("helper", func() {
 		})
 	})
 
+	Describe("#MonitoringConfiguration", func() {
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/helper/helpers_test.go]:
```suggestion
	Describe("#IsMonitoringEnabled", func() {
```

File: pkg/gardenlet/apis/config/helper/helpers_test.go [hunk 10]
```diff
@@ -121,6 +121,35 @@ var _ = Describe("helper", func() {
...
+				Monitoring: &config.MonitoringConfig{
+					Shoot: &config.ShootMonitoringConfig{
+						Enabled: pointer.Bool(true),
+					},
+				},
+			}
+			Expect(IsMonitoringEnabled(gardenletConfig)).To(BeTrue())
+		})
+
+		It("should return true when nothing is set", func() {
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/helper/helpers_test.go:147]: Thanks for adding the tests. Can you add a few more for when `c.Monitoring.Shoot = nil` and `c.Monitoring.Shoot.Enabled = nil`

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 11]
```diff
@@ -681,6 +681,10 @@ type MonitoringConfig struct {
 
 // ShootMonitoringConfig contains settings for the shoot monitoring stack.
 type ShootMonitoringConfig struct {
+	// Enabled is used to enable or disable the shoot monitoring stack.
+	// Defaults to true.
+	//+optional
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/v1alpha1/types.go]:
```suggestion
	// +optional
```

File: pkg/operation/botanist/monitoring.go (hunk 2 above)
Comment (plkokanov) [reply pkg/operation/botanist/monitoring.go]: cc @nickytd @vlvasilev @Kristian-ZH we should keep `grafana` if only the monitoring component is disabled right?

**Review by plkokanov — CHANGES_REQUESTED** [review]
Could you check why `make check-generate` is failing. I think that there shouldn't be any changes in `charts/gardener/provider-local/registration/templates/controller-registration.yaml` and `example/provider-local/garden/base/controller-registration.yaml`

Additionally, could you provide a bit more detailed description on why we are making the monitoring component optional?

File: pkg/gardenlet/apis/config/helper/helpers_test.go (hunk 9 above)
Comment (nschad) [reply pkg/gardenlet/apis/config/helper/helpers_test.go]: why? it's the same pattern as
```go
	Describe("#LoggingConfiguration", func() {
```

**Comment by nschad** [comment]
> Could you check why `make check-generate` is failing. I think that there shouldn't be any changes in `charts/gardener/provider-local/registration/templates/controller-registration.yaml` and `example/provider-local/garden/base/controller-registration.yaml`
> 
> Additionally, could you provide a bit more detailed description on why we are making the monitoring component optional?

No Idea. Whenever I run `make generate` this is what I get.

You can find the entire discussion here: #5650

File: pkg/gardenlet/apis/config/helper/helpers_test.go (hunk 10 above)
Comment (nschad) [reply pkg/gardenlet/apis/config/helper/helpers_test.go:147]: done!

File: pkg/operation/botanist/monitoring.go (hunk 2 above)
Comment (Kristian-ZH) [reply pkg/operation/botanist/monitoring.go]: Yes, the logging stack uses Grafana too.
You can inert the additional check in your `if`: `gardenlethelper.IsLokiEnabled()`

Comment (nschad) [reply pkg/operation/botanist/monitoring.go]: Added loki to the check
```go
	if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting ||
		(!gardenlethelper.IsMonitoringEnabled(b.Config) && !gardenlethelper.IsLokiEnabled(b.Config)) {
```

**Comment by nschad** [comment]
Do i need to act on the failing `pull-gardener-apidiff` test? Honestly no Idea what I'm supposed to do here

**Comment by shafeeqes** [comment]
> Do i need to act on the failing `pull-gardener-apidiff` test? Honestly no Idea what I'm supposed to do here

It's not a "required" test to merge.

File: docs/monitoring/README.md [hunk 12]
```diff
@@ -69,4 +69,12 @@ monitoring:
...
\ No newline at end of file
+If basic auth is needed it can be set via secret in garden namespace (Gardener API Server). [Example secret](../../example/10-secret-remote-write.yaml)
+
+## Disable Gardener Monitoring
+
+If you wish to disable metric collection for every shoot and roll your own then you can simply set.
+```
+monitoring:
+  enabled: false
+```
```
Comment (timebertt) [inline docs/monitoring/README.md]:
```suggestion
If you wish to disable metric collection for every shoot and roll your own then you can simply set:
```yaml
monitoring:
  enabled: false
```

File: example/20-componentconfig-gardenlet.yaml [hunk 13]
```diff
@@ -192,6 +192,7 @@ etcdConfig:
 #   reelectionPeriod: 5s
 #   etcdConnectionTimeout: 5s
 # monitoring:
+#   enabled: true
```
Comment (timebertt) [inline example/20-componentconfig-gardenlet.yaml]: I think this needs to go below `shoot:`, correct?

File: pkg/gardenlet/apis/config/helper/helpers.go [hunk 14]
```diff
@@ -109,3 +109,12 @@ func IsLokiEnabled(c *config.GardenletConfiguration) bool {
 	}
 	return true
 }
+
+// IsMonitoringEnabled returns true if the monitoring stack for shoot clusters is enabled. Default is enabled.
+func IsMonitoringEnabled(c *config.GardenletConfiguration) bool {
+	if c != nil && c.Monitoring != nil && c.Monitoring.Shoot != nil &&
```
Comment (timebertt) [inline pkg/gardenlet/apis/config/helper/helpers.go:115]: The monitoring stack for shoots should be enabled by default (=previous behavior).
Can you invert the logic to the following?
```go
return c == nil || c.Monitoring == nil || c.Monitoring.Shoot == nil || c.Monitoring.Shoot.Enabled == nil || *c.Monitoring.Shoot.Enabled
```

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 15]
```diff
@@ -681,6 +681,10 @@ type MonitoringConfig struct {
 
 // ShootMonitoringConfig contains settings for the shoot monitoring stack.
 type ShootMonitoringConfig struct {
+	// Enabled is used to enable or disable the shoot monitoring stack.
+	// Defaults to true.
+	// +optional
+	Enabled *bool `json:"enabled,omitempty"`
```
Comment (timebertt) [inline pkg/gardenlet/apis/config/v1alpha1/types.go:687]: This field should be defaulted explicitly to `true` in `defaults.go`

**Review by timebertt — CHANGES_REQUESTED** [review]
Thanks @nschad for the PR!
I have added some comments from a quick scan. Will check the remaining parts as well :)

File: example/20-componentconfig-gardenlet.yaml (hunk 13 above)
Comment (plkokanov) [reply example/20-componentconfig-gardenlet.yaml:195]: Since this is only used to disable the monitoring for shoot and will not affect the monitoring stack for seeds that is deployed in the `garden` namespace I would also vote to make the option: `monitoring.shoot.enabled`

File: pkg/gardenlet/apis/config/helper/helpers.go (hunk 14 above)
Comment (nschad) [reply pkg/gardenlet/apis/config/helper/helpers.go:115]: > The monitoring stack for shoots should be enabled by default (=previous behavior).

The default is `true`? We basically did the same as `IsLokiEnabled` and `IsLoggingEnabled` However I can use your code-snippet if you want.

**Comment by nschad** [comment]
I've re-added
>   charts/gardener/provider-local/registration/templates/controller-registration.yaml
>   example/provider-local/garden/base/controller-registration.yaml

I feel like somebody forgot it in an earlier PR otherwise no idea why `make generate` always creates/updates these files for me.

File: pkg/gardenlet/apis/config/v1alpha1/types.go (hunk 15 above)
Comment (nschad) [reply pkg/gardenlet/apis/config/v1alpha1/types.go:687]: done! I hope I did that correctly

Comment (plkokanov) [reply pkg/gardenlet/apis/config/v1alpha1/types.go:687]: You also need to set `GardenletConfiguration.Monitoring` and `GardenletConfiguration.Monitoring.Shoot` to not be nil. You can check how it's done for `GardenletConfiguration.SNI` and `GardenletConfiguration.SNI.Ingress`:  https://github.com/gardener/gardener/blob/c29fe5452bdd8f81e5d40256a8325df9a16ca688/pkg/gardenlet/apis/config/v1alpha1/defaults.go#L81-L83
https://github.com/gardener/gardener/blob/c29fe5452bdd8f81e5d40256a8325df9a16ca688/pkg/gardenlet/apis/config/v1alpha1/defaults.go#L444-L472

And you would also need to add tests in `defaults_test.go`

**Comment by plkokanov** [comment]
> I've re-added
> 
> > charts/gardener/provider-local/registration/templates/controller-registration.yaml
> > example/provider-local/garden/base/controller-registration.yaml
> 
> I feel like somebody forgot it in an earlier PR otherwise no idea why `make generate` always creates/updates these files for me.

Hmm, I checked out your PR before you re-added the charts to test it locally and ran `make generate` but the charts did not get regenerated for me. What `go` version are you using?

File: pkg/gardenlet/apis/config/helper/helpers.go (hunk 14 above)
Comment (timebertt) [reply pkg/gardenlet/apis/config/helper/helpers.go:115]: You're right. Stupid me, forget what I said 🙈

**Comment by nschad** [comment]
> > I've re-added
> > > charts/gardener/provider-local/registration/templates/controller-registration.yaml
> > > example/provider-local/garden/base/controller-registration.yaml
> > 
> > 
> > I feel like somebody forgot it in an earlier PR otherwise no idea why `make generate` always creates/updates these files for me.
> 
> Hmm, I checked out your PR before you re-added the charts to test it locally and ran `make generate` but the charts did not get regenerated for me. What `go` version are you using?

go version go1.18.3 linux/amd64

**Comment by nschad** [comment]
@plkokanov So i tested the whole `make generate` again with the following:

> * rm -rf gardener/
> * fresh git clone
> * git checkout c29fe5452bdd8f81e5d40256a8325df9a16ca688 (latest upstream master commit)
> * make generate
> * and swoop....it generated the controller registration files.

Also upgraded go from 1.18.3 to 1.18.4. I'll ask a colleague to confirm my findings.

**Comment by nschad** [comment]
@plkokanov Okay something seems off with either my helm, tar, gzip (even though its default (GNU) ubuntu). I have no clue at the moment, cannot reproduce on a different system. I will just checkout the files from upstream/master 
`¯\_(ツ)_/¯`

**Comment by plkokanov** [comment]
Generally, the correct `helm` binary should be downloaded and used as part of the `make` command. We make use of a specials `tools.mk` file where such tool dependencies are listed. However, I think we don't include `tar` `gzip` and `sed` all of which are used for the local provider controller-registration.yaml generation. So indeed there could be some mismatch in those tools.

File: pkg/gardenlet/apis/config/v1alpha1/defaults.go [hunk 16]
```diff
@@ -211,6 +215,21 @@ func SetDefaults_BackupEntryControllerConfiguration(obj *BackupEntryControllerCo
 	}
 }
 
+// SetDefaults_MonitoringConfig sets the defaults for the shoot monitoring stack
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/v1alpha1/defaults.go]:
```suggestion
// SetDefaults_MonitoringConfig sets the defaults for the monitoring stack.
```

File: pkg/gardenlet/apis/config/v1alpha1/defaults.go [hunk 17]
```diff
@@ -211,6 +215,21 @@ func SetDefaults_BackupEntryControllerConfiguration(obj *BackupEntryControllerCo
...
+func SetDefaults_MonitoringConfig(obj *MonitoringConfig) {
+	if obj.Shoot == nil {
+		obj.Shoot = &ShootMonitoringConfig{}
+	}
+}
+
+// SetDefaults_ShootMonitoringConfig sets the defaults for the shoot monitoring
```
Comment (plkokanov) [inline pkg/gardenlet/apis/config/v1alpha1/defaults.go]:
```suggestion
// SetDefaults_ShootMonitoringConfig sets the defaults for the shoot monitoring.
```

File: pkg/operation/botanist/addons.go [hunk 18]
```diff
@@ -235,13 +237,18 @@ func (b *Botanist) generateCoreAddonsChart(ctx context.Context) (*chartrenderer.
 		return nil, err
 	}
 
+	monitoringEnabled := helper.IsMonitoringEnabled(b.Config)
```
Comment (plkokanov) [inline pkg/operation/botanist/addons.go]:
```suggestion
	monitoringEnabled := helper.IsMonitoringEnabled(b.Config) && b.Shoot.Purpose != gardencorev1beta1.ShootPurposeTesting
```
Just a suggestion to combine this and lines 241-243

**Review by plkokanov — CHANGES_REQUESTED** [review]
Just a few nits left. Rest looks good I think.

**Comment by nschad** [comment]
> Just a few nits left. Rest looks good I think.

Done.

File: pkg/operation/botanist/addons.go [hunk 19]
```diff
@@ -235,13 +237,15 @@ func (b *Botanist) generateCoreAddonsChart(ctx context.Context) (*chartrenderer.
 		return nil, err
 	}
 
+	monitoringEnabled := helper.IsMonitoringEnabled(b.Config) && b.Shoot.Purpose != gardencorev1beta1.ShootPurposeTesting
```
Comment (timebertt) [inline pkg/operation/botanist/addons.go]: Now we have a lot of duplicated checks like this: "is monitoring enabled in gardenlet config" and "is shoot not of type purpose"?
This is likely to be regressed by future changes.

Can you deduplicate them by adding a function like `IsShootMonitoringEnabled` on `*Operation`?

File: pkg/operation/care/checker.go [hunk 20]
```diff
@@ -475,12 +475,13 @@ func (b *HealthChecker) CheckClusterNodes(
 func (b *HealthChecker) CheckMonitoringControlPlane(
 	namespace string,
 	isTestingShoot bool,
+	monitoringEnabled bool,
```
Comment (timebertt) [inline pkg/operation/care/checker.go]: Can you combine these two parameters into one `shootMonitoringEnabled` parameter?

File: pkg/operation/botanist/monitoring.go [hunk 21]
```diff
@@ -424,7 +424,9 @@ func (b *Botanist) DeploySeedMonitoring(ctx context.Context) error {
 
 // DeploySeedGrafana deploys the grafana charts to the Seed cluster.
 func (b *Botanist) DeploySeedGrafana(ctx context.Context) error {
-	if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting {
+	// disable monitoring if shoot has purpose testing or monitoring and loki is disabled
+	if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting ||
+		(!gardenlethelper.IsMonitoringEnabled(b.Config) && !gardenlethelper.IsLokiEnabled(b.Config)) {
```
Comment (timebertt) [inline pkg/operation/botanist/monitoring.go]: Same here: can we combine this into a single reusable check `Operation.WantsGrafana` or something similar?

**Review by timebertt — COMMENTED** [review]
Some last nit comments regarding style.
The rest looks good, also verified locally :)

File: pkg/operation/botanist/addons.go (hunk 19 above)
Comment (nschad) [reply pkg/operation/botanist/addons.go]: Is there some example on how to do this? I can't really follow

Comment (nschad) [reply pkg/operation/botanist/addons.go]: ah I think I've found it. Hold on :D

**Comment by nschad** [comment]
Okay phew :P

Let me know if that works for you @timebertt

File: pkg/operation/operation.go [hunk 22]
```diff
@@ -608,6 +609,16 @@ func (o *Operation) ComputeLokiHosts() []string {
...
+// IsShootMonitoringEnabled returns true if shoot monitoring is enabled and shoot is not of purpose testing.
+func (o *Operation) IsShootMonitoringEnabled() bool {
+	return helper.IsMonitoringEnabled(o.Config) && o.Shoot.Purpose != gardencorev1beta1.ShootPurposeTesting
+}
+
+// WantsGrafana returns true if shoot is not of purpose testing and either shoot monitoring or loki is enabled.
+func (o *Operation) WantsGrafana() bool {
+	return o.Shoot.Purpose != gardencorev1beta1.ShootPurposeTesting && (helper.IsMonitoringEnabled(o.Config) || helper.IsLokiEnabled(o.Config))
+}
+
```
Comment (shafeeqes) [inline pkg/operation/operation.go:622]: You could add unit tests.

File: pkg/operation/botanist/monitoring.go [hunk 23]
```diff
@@ -69,7 +69,7 @@ func observabilityIngressSecretConfig(name string) *secrets.BasicAuthSecretConfi
 // DeploySeedMonitoring installs the Helm release "seed-monitoring" in the Seed clusters. It comprises components
 // to monitor the Shoot cluster whose control plane runs in the Seed cluster.
 func (b *Botanist) DeploySeedMonitoring(ctx context.Context) error {
-	if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting {
+	if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting || !gardenlethelper.IsMonitoringEnabled(b.Config) {
```
Comment (timebertt) [inline pkg/operation/botanist/monitoring.go]: You can use your new function here as well :)
```suggestion
	if !b.IsShootMonitoringEnabled() {
```

File: pkg/operation/operation.go (hunk 22 above)
Comment (nschad) [reply pkg/operation/operation.go:622]: No. I'm not gonna unit test a boolean check. The helper functions are already tested.

File: pkg/operation/operation.go [hunk 24]
```diff
@@ -608,6 +610,16 @@ func (o *Operation) ComputeLokiHosts() []string {
 	}
 }
 
+// IsShootMonitoringEnabled returns true if shoot monitoring is enabled and shoot is not of purpose testing.
+func (o *Operation) IsShootMonitoringEnabled() bool {
```
Comment (timebertt) [inline pkg/operation/operation.go:614]: I'm wondering: should we also use this in `kubestatemetrics.go`?
Otherwise, `kube-state-metrics` will get deployed even if shoot monitoring is disabled.

IIRC, `kube-state-metrics` is scraped by the seed prometheus. Hence, I'm not sure if it's desired to keep `kube-state-metrics`. 
Do you want to keep it in your landscapes?

```diff
diff --git a/pkg/operation/botanist/kubestatemetrics.go b/pkg/operation/botanist/kubestatemetrics.go
index 313eddd92..5e540e114 100644
--- a/pkg/operation/botanist/kubestatemetrics.go
+++ b/pkg/operation/botanist/kubestatemetrics.go
@@ -17,7 +17,6 @@ package botanist
 import (
        "context"

-       gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
        "github.com/gardener/gardener/pkg/operation/botanist/component"
        "github.com/gardener/gardener/pkg/operation/botanist/component/kubestatemetrics"
        "github.com/gardener/gardener/pkg/utils/images"
@@ -45,7 +44,7 @@ func (b *Botanist) DefaultKubeStateMetrics() (kubestatemetrics.Interface, error)

 // DeployKubeStateMetrics deploys or destroys the kube-state-metrics to the shoot namespace in the seed.
 func (b *Botanist) DeployKubeStateMetrics(ctx context.Context) error {
-       if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting {
+       if !b.IsShootMonitoringEnabled() {
                return b.Shoot.Components.ControlPlane.KubeStateMetrics.Destroy(ctx)
        }

```

**Review by timebertt — COMMENTED** [review]
> Let me know if that works for you @timebertt

Yes, this is what I was looking for 👍 
One nit and one question :)

**Comment by timebertt** [comment]
@nschad FYI: you don't need to squash your commits, prow will do this for you when merging your PR :) 
So, feel free to add as many commits as you wish 😄

File: pkg/operation/operation.go (hunk 24 above)
Comment (timebertt) [reply pkg/operation/operation.go:614]: > IIRC, `kube-state-metrics` is scraped by the seed prometheus. 

Ok, I double-checked this and it seems I didn't remember correctly.
`garden/prometheus` scrapes `garden/kube-state-metrics` including metrics for shoot namespaces.
`garden/seed-prometheus` scrapes `garden/kube-state-metrics` but drops metrics for shoot control plane namespaces.
`shoot--foo--bar/prometheus` scrapes `garden/kube-state-metrics` but only keeps metrics for the shoot control plane namespace (`shoot--foo--bar`).
`shoot--foo--bar/prometheus` scrapes `shoot--foo--bar/kube-state-metrics` which includes metrics for objects in the shoot cluster itself.

TL;DR: I think the `kube-state-metrics` for the shoot cluster itself should be disabled as well if the shoot monitoring is disabled.
I.e., you can simply apply the proposed patch from my last comment :)

Comment (timebertt) [reply pkg/operation/operation.go:614]: Sorry for the back and forth 🙊

Comment (nschad) [reply pkg/operation/operation.go:614]: nice catch. Our older gardener version didn't had that yet `ಠ_ಠ`

**Review by timebertt — APPROVED** [review]
Awesome, thanks for addressing my suggestions. Let's get this in!
/lgtm
